gridcomps/History3G/OutputInfo.F90

```diff
@@ -0,0 +1,331 @@
+      _RETURN_UNLESS(n > 0)
+      call ESMF_InfoGetCharAlloc(info, key=KEY_VLOC, value=spec_name, _RC)
+      _RETURN(_SUCCESS)
+
+   end function get_vertical_dim_spec_info
+
+   function get_ungridded_dims_bundle(bundle, rc) result(dims)
+      type(UngriddedDims) :: dims
+      type(ESMF_FieldBundle), intent(in) :: bundle
+      integer, optional, intent(out) :: rc
+      integer :: status
+      integer :: i
+      type(ESMF_Info), allocatable :: info(:)
+      type(UngriddedDimVector) :: vec
+
+      info = get_bundle_info(bundle, _RC)
+      vec = get_ungridded_dims_bundle_info(info, _RC)
+      dims = UngriddedDims(vec)
+      call destroy_bundle_info(info, _RC)
+      _RETURN(_SUCCESS)
+
+   end function get_ungridded_dims_bundle
+
+   function get_ungridded_dims_bundle_info(info, rc) result(vec)
+      type(UngriddedDimVector) :: vec
+      type(ESMF_Info), intent(in) :: info(:)
+      integer, optional, intent(out) :: rc
+      integer :: status
+      integer :: i
+
+      vec = UngriddedDimVector()
+      do i=1, size(info)
+         call push_ungridded_dim(vec, info(i), _RC)
+      end do
+      _RETURN(_SUCCESS)
+
+   end function get_ungridded_dims_bundle_info
+
+   function get_ungridded_dims_field(field, rc) result(ungridded)
+      type(UngriddedDims) :: ungridded
+      type(ESMF_Field), intent(inout) :: field
+      integer, optional, intent(out) :: rc
+      integer :: status
+      type(ESMF_Info) :: info
+      type(UngriddedDimVector) :: vec
+
+      call ESMF_InfoGetFromHost(field, info, _RC)
+      call push_ungridded_dim(vec, info, _RC)
+      ungridded = UngriddedDims(vec)
+      call ESMF_InfoDestroy(info, _RC)
+      _RETURN(_SUCCESS)
+
+   end function get_ungridded_dims_field
+
+   subroutine push_ungridded_dim(vec, info, rc)
+      type(UngriddedDimVector), intent(inout) :: vec
+      type(ESMF_Info), intent(in) :: info
+      integer, optional, intent(out) :: rc 
+      integer :: status
+      type(UngriddedDim) :: next
+      integer :: num_dims, i, vi
+      logical :: has_dims
+      integer :: num_coord
+      character(len=:), allocatable :: name
+      character(len=:), allocatable :: units
+      character(len=:), allocatable :: dim_key
+      real, allocatable :: coordinates(:)
+
+      num_dims = 0
+      has_dims = ESMF_InfoIsPresent(info, key=KEY_NUM_UNGRID_DIMS, _RC)
+      if(has_dims) then
+         call ESMF_InfoGet(info, key=KEY_NUM_UNGRID_DIMS, value=num_dims, _RC)
+      end if
+      do i=1, num_dims
+         dim_key = make_dim_key(i, _RC)
+         call ESMF_InfoGetCharAlloc(info, key=dim_key // KEY_UNGRIDDED_NAME, value=name, _RC)
+         call ESMF_InfoGetCharAlloc(info, key=dim_key // KEY_UNGRIDDED_UNITS, value=units, _RC)
+         call ESMF_InfoGet(info, key=dim_key // KEY_UNGRIDDED_COORD, size=num_coord, _RC)
+         SAFE_ALLOC1(coordinates, num_coord)
+         call ESMF_InfoGet(info, key=dim_key // KEY_UNGRIDDED_COORD, values=coordinates, _RC)
```

Comment:
Would be nice if ESMF did the allocation here.  Sigh.

Reply:
Oh - and then I remembered that it can:   `ESMF_InfoGetAlloc`

You probably then don't need the macros and safeguards.  Allocate once and deallocate once in each iteration.  (With another change I'm going to propose.)

Reply:
done